tcf_website/static/department/sorting.js

```diff
@@ -0,0 +1,216 @@
+    }
+}
+
+const sort_by_difficulty = () => {
+    if (!$('#diff-sort-btn').hasClass('active')) {
+        $('#number-sort-btn').removeClass('active')
+        $('#rating-sort-btn').removeClass('active')
+        $('#gpa-sort-btn').removeClass('active')
+        $('#diff-sort-btn').addClass('active')
+    } else {
+        difficulty_A = !difficulty_A;
+    }
+
+    if (difficulty_A) {
+        sort_classes(cmp_by_difficulty_A)
+    } else {
+        sort_classes(cmp_by_difficulty_D)
+    }
+}
+
+const sort_by_gpa = () => {
+    if (!$('#gpa-sort-btn').hasClass('active')) {
+        $('#number-sort-btn').removeClass('active')
+        $('#rating-sort-btn').removeClass('active')
+        $('#diff-sort-btn').removeClass('active')
+        $('#gpa-sort-btn').addClass('active')
+    } else {
+        gpa_A = !gpa_A;
+    }
+
+    if (gpa_A) {
+        sort_classes(cmp_by_gpa_A)
+    } else {
+        sort_classes(cmp_by_gpa_D)
+    }
+}
+
+
+const sort_classes = (cmp) => {
+    $("ul.course-list").each(function (a) {
+        var li = $(this).children("li");
+        li.detach().sort(cmp);
+        $(this).append(li);
+    });
+}
+
+const cmp_by_rating_D = (a, b) => {
+    ratingA = a.querySelectorAll(".mb-0.info")[0].innerHTML.trim();
+    ratingB = b.querySelectorAll(".mb-0.info")[0].innerHTML.trim();
+
+    if (ratingA === '—') {
+        return 1;
+    } else if (ratingB === '—') {
+        return -1;
+    }
+
+    if (ratingA > ratingB) {
+        return -1;
+    } else if (ratingB > ratingA) {
+        return 1;
+    } else {
+        return 0;
+    }
+};
+
+const cmp_by_rating_A = (a, b) => {
+    ratingA = a.querySelectorAll(".mb-0.info")[0].innerHTML.trim();
+    ratingB = b.querySelectorAll(".mb-0.info")[0].innerHTML.trim();
+
+    if (ratingA === '—') {
+        return 1;
+    } else if (ratingB === '—') {
+        return -1;
+    }
+
+    if (ratingA > ratingB) {
+        return 1;
+    } else if (ratingB > ratingA) {
+        return -1;
+    } else {
+        return 0;
+    }
+};
+
+const cmp_by_difficulty_D = (a, b) => {
+    difficultyA = a.querySelectorAll(".mb-0.info")[1].innerHTML.trim();
+    difficultyB = b.querySelectorAll(".mb-0.info")[1].innerHTML.trim();
+
+    if (difficultyA === '—') {
+        return 1;
+    } else if (difficultyB === '—') {
+        return -1;
+    }
+
+    if (difficultyA > difficultyB) {
+        return -1;
+    } else if (difficultyB > difficultyA) {
+        return 1;
+    } else {
+        return 0;
+    }
+};
+
+const cmp_by_difficulty_A = (a, b) => {
+    difficultyA = a.querySelectorAll(".mb-0.info")[1].innerHTML.trim();
+    difficultyB = b.querySelectorAll(".mb-0.info")[1].innerHTML.trim();
+
+    if (difficultyA === '—') {
+        return 1;
+    } else if (difficultyB === '—') {
+        return -1;
+    }
+
+    if (difficultyA > difficultyB) {
+        return 1;
+    } else if (difficultyB > difficultyA) {
+        return -1;
+    } else {
+        return 0;
+    }
+};
```

Comment:
Is there a way to condense these into one function? Like have one function that can do both? It's OK if that's not easy to do-- then we'll just stick with having 2.

Reply:
I was thinking about this and I wasn't sure if it was possible. It's not possible to simply reverse the array as I want the NaN elements to be at the bottom regardless of asc/desc. I'm also using the Array.sort() method which can take a comparator function but it needs to have two parameters only. I'm not sure if there's a way to implement a partial function of sorts to condense the asc/desc and maybe all the comparator functions as well.

Reply:
Potentially (not sure if this is worth), you could filter out NaN elements, sort that, reverse if needed, and then concat with NaN elements. Might not be worth the additional complexity.

Another idea would be to 
```
const compare_by_difficulty_partial(ascending) {

    const compare_by_difficulty = (a, b) => {
        const difficultyA = a.querySelectorAll(".mb-0.info")[1].innerHTML.trim();
        const difficultyB = b.querySelectorAll(".mb-0.info")[1].innerHTML.trim();

        if (difficultyA === '—') {
            return 1;
        } else if (difficultyB === '—') {
            return -1;
        }

        if (difficultyA > difficultyB) {
            return Number(ascending);
        } else if (difficultyB > difficultyA) {
            return -1;
        } else {
            return Number(!ascending);
        }
    }

    return compare_by_difficulty;
}
```

Reply:
Also, should we be using `parseFloat()` here?

Reply:
I was simply doing string comparisons here as the ordering would be the same. I could convert to `parseFloat()` if that's better? Also, I think I might lean towards the partial implementation as I could then condense all of the functions into one.

Reply:
You're right, string comparison is fine here. Maybe add a comment so that future readers won't think it's a bug?